### PR TITLE
GSASIIRefineFitPeaks reintegrated with latest version of GSAS-II

### LINKED ISSUE #19578: Engg - Reintegrate GSAS-II with existing Mantid algorithms
As part of the ongoing engineering work we need to integrate GSAS-II with Mantid. We have an existing algorithm `GSASIIRefineFitPeaks` however this relies on the internals of an older copy of GSAS-II

### PR DESCRIPTION
**Pull request by josephframsay**
The algorithm `GSASIIRefineFitPeaks`  has been brought up to date and integrated with the most recent version of GSAS-II. The single peak fitting functionality has been disabled for the time being, as the new scriptable API for GSAS-II does not support it.

More functionality is expected to be added to the algorithm as scientists' specific fitting needs become apparent, but this is a good starting block for bringing support for engineering back into Mantid.

**To test:**

You'll need a fairly recent version of GSAS-II (later than April 2017) - you can get it [here](https://subversion.xray.aps.anl.gov/trac/pyGSAS). Some example data is [here](https://github.com/mantidproject/mantid/files/1397963/GSASIIRefineFitPeaks_data.zip)

The following script, as a minimum should give you some data:
```
ws = Load(Filename=<path to data>/"focused_bank1_ENGINX00256663.nxs")

GOF, Rwp, lattice_tbl = GSASIIRefineFitPeaks(PathToGSASII=<path to gsas>, 
                                             SaveGSASIIProjectFile=<some new GSAS-II project>,
                                             PhaseInfoFile=<path to dath>/"FE_ALPHA.cif",
                                             InstrumentFile=<path to data>/"template_ENGINX_241391_236516_North_bank.prm">,
                                             InputWorkspace=ws,
                                             RefinementMethod="Rietveld refinement")                                                          
print "Goodness of fit coefficient: {0:.5f}".format(GOF)
print "Weighted profile R-factor (Rwp): {0:.5f}".format(Rwp)
print ("Lattice parameters, a: {a}, b: {b}, c: {c}, alpha: {alpha}, beta: {beta}, gamma: {gamma}, "
         "Volume: {volume:.3f}".format(**lattice_tbl.row(0)))
```

Try setting `RefinementMethod="Pawley refinement"`. The unit tests won't run on the build servers, as they are skipped if GSAS-II is not found, so try running those on a local machine too (you need to have GSAS-II in a directory called `g2conda` (the installation default) within 2 levels of your root directory.

Also code review, as the algorithm has been completely rewritten and will be added to, so I want to make sure I haven't shot myself (or anyone else) in the foot without noticing early on.

Fixes #19578

**Release Notes** 

[Here](https://github.com/mantidproject/mantid/commit/bb3ce4705b2dd04bfa5247de41b20015839f44d9)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by samueljackson92** [comment]
@josephframsay Can you tell me exactly what revision you used with this?

File: Framework/PythonInterface/test/python/plugins/algorithms/GSASIIRefineFitPeaksTest.py [hunk 1]
```diff
@@ -1,67 +1,136 @@
...
+            return cur_dir_path
+
+        list_dir = os.listdir(cur_dir_path)
+        for child in list_dir:
+            child_path = os.path.join(cur_dir_path, child)
+            if os.path.isdir(child_path):
+                try:
+                    path = _GSASFinder._find_directory_by_name(cur_dir_name=child, cur_dir_path=child_path,
+                                                               level=level + 1, name_to_find=name_to_find,
+                                                               max_level=max_level)
```
Comment (samueljackson92) [inline Framework/PythonInterface/test/python/plugins/algorithms/GSASIIRefineFitPeaksTest.py]: Any reason we can't use `os.walk` here instead?

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py [hunk 2]
```diff
@@ -15,974 +16,227 @@ class GSASIIRefineFitPeaks(PythonAlgorithm):
...
-        self.PROP_OUT_GOF = 'GoF'
-        self.PROP_OUT_RWP = 'Rwp'
-        self.PROP_REFINE_CENTER = 'RefineCenter'
-        self.PROP_REFINE_INTENSITY = 'RefineIntensity'
-        self.PROP_REFINE_ALPHA = 'RefineAlpha'
-        self.PROP_REFINE_BETA = 'RefineBeta'
-        self.PROP_REFINE_SIGMA = 'RefineSigma'
-        self.PROP_REFINE_GAMMA = 'RefineGamma'
-        self.PROP_METHOD = "Method"
+        return ("Perform Rietveld refinement of lattice parameters on a diffraction spectrum "
```
Comment (samueljackson92) [inline Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: This allows for more than just rietveld, as we can also do pawley refinement

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py [hunk 3]
```diff
@@ -15,974 +16,227 @@ class GSASIIRefineFitPeaks(PythonAlgorithm):
...
+                                                     do_pawley=refinement_method == self.REFINEMENT_METHODS[0])
+            self._set_output_properties(rwp=rwp, gof=gof, lattice_params=lattice_params)
 
-        limits = self._build_add_limits(gs2_rd)
-        self.log().information("Fitting loaded histogram data, with limits: {0}".format(limits))
-
-        # self.PROP_PHASE_INFO_FILE - phase information into rd, this is loaded elsewhere when
-        # doing Pawley/Rietveld refinement
+    def _build_output_lattice_table(self, lattice_params):
+        alg = self.createChildAlgorithm('CreateEmptyTableWorkspace')
```
Comment (samueljackson92) [inline Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: Does this need the `StoreInADS=False` thing now?

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py [hunk 4]
```diff
@@ -15,974 +16,227 @@ class GSASIIRefineFitPeaks(PythonAlgorithm):
...
-
-            # Flag for Pawley intensity extraction (bool)
-            general_phase_data['doPawley'] = True
-            # maximum Q (as d-space) to use for Pawley extraction
-            general_phase_data['Pawley dmin'] = self.getProperty(self.PROP_PAWLEY_DMIN).value
-            # Restraint value for negative Pawley intensities
-            general_phase_data['Pawley neg wt'] = self.getProperty(self.PROP_PAWLEY_NEG_WEIGHT).value
+        gsas_path = self.getPropertyValue(self.PROP_PATH_TO_GSASII)
+        sys.path.append(gsas_path)
+        import GSASIIscriptable as GSASII
```
Comment (samueljackson92) [inline Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: This should be in a try/except and we need to report a very loud error to the log and the script window that GSAS is not installed and could not be found.

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py [hunk 5]
```diff
@@ -15,974 +16,227 @@ class GSASIIRefineFitPeaks(PythonAlgorithm):
...
+
+        pawley_neg_wt = self.getPropertyValue(self.PROP_PAWLEY_NEGATIVE_WEIGHT)
+        phase_params["Pawley neg wt"] = pawley_neg_wt
+
+    @contextmanager
+    def _suppress_stdout(self):
+        """
+        Suppress output from print statements. This is mainly useful for debugging, as GSAS does a lot of printing.
+        """
+        if self.getPropertyValue(self.PROP_SUPPRESS_GSAS_OUTPUT) == "1":
```
Comment (samueljackson92) [inline Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: This is a boolean property. Can we not get it as a boolean?

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Generally good. A few further changes required.

File: Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py (hunk 3 above)
Comment (josephframsay) [reply Framework/PythonInterface/plugins/algorithms/GSASIIRefineFitPeaks.py]: I don't think so - child algorithms don't store in the ADS anyway, right? I thought that `StoreInADS` was for calling algorithms through `simpleapi`

File: Framework/PythonInterface/test/python/plugins/algorithms/GSASIIRefineFitPeaksTest.py (hunk 1 above)
Comment (josephframsay) [reply Framework/PythonInterface/test/python/plugins/algorithms/GSASIIRefineFitPeaksTest.py]: I could be wrong, but don't think `os.walk` can be depth-limited (you can filter the results by depth, but you still have to go through the entire iterator), and on Linux at least does a depth-first search, so it'll fail on the build servers due to timeout

Comment (samueljackson92) [reply Framework/PythonInterface/test/python/plugins/algorithms/GSASIIRefineFitPeaksTest.py]: In that case this is probably fine.

**Review by samueljackson92 — APPROVED** [review]

**Pull request merged**